pkg/autoscaling/calculation.go

```diff
@@ -52,89 +49,399 @@ var (
 	}
 
-	return groups
+	return nil, nil
+}
+
+func getHeterogeneousScaleOutPlans(cpuScaleOutSize float64, cpuUsageTarget float64, availableCount uint64, resourceMap map[string]uint64, resources []*Resource, component ComponentType) []*Plan {
+	var plans []*Plan
+	// sort resources by cpu desc
+	sortResourcesByCPUDesc(resources)
+
+	for _, resource := range resources {
+		if cpuScaleOutSize <= 0 || availableCount <= 0 {
+			break
+		}
+
+		if resource.ResourceType != homogeneousTiKVResourceType && resource.ResourceType != homogeneousTiDBResourceType {
+			scaleOutCount := uint64(cpuScaleOutSize/float64(resource.CPU)/cpuUsageTarget) + 1
+			if scaleOutCount > availableCount {
+				// not enough k8s nodes to scale out, reduce the scale out count
+				scaleOutCount = availableCount
+				availableCount = 0
+			}
+
+			existsCount, ok := resourceMap[resource.ResourceType]
+			if ok {
+				// this resource type exists
+				count := scaleOutCount + existsCount
+				if resource.Count == nil || count <= *resource.Count {
+					// unlimited resource count or enough resource count left
+					plans = append(plans, NewPlan(component, scaleOutCount, resource.ResourceType))
+					log.Debug("get heterogeneous scale out plans completed",
+						zap.String("component", component.String()),
+						zap.Any("plans", plans),
+					)
+
+					return plans
+				}
+
+				// not enough count left, use as much as possible
+				scaleOutCount = *resource.Count - existsCount
+				availableCount -= scaleOutCount
+				count = *resource.Count
+				cpuScaleOutSize -= float64(resource.CPU * scaleOutCount)
+
+				plans = append(plans, NewPlan(component, count, resource.ResourceType))
+				continue
+			}
+
+			// this resource type does not exist
+			if resource.Count == nil || scaleOutCount <= *resource.Count {
+				// unlimited resource count or enough resource count left
+				plans = append(plans, NewPlan(component, scaleOutCount, resource.ResourceType))
+				log.Debug("get heterogeneous scale out plans completed",
+					zap.String("component", component.String()),
+					zap.Any("plans", plans),
+				)
+
+				return plans
+			}
+
+			if *resource.Count > 0 {
+				// not enough count left, use as much as possible
+				availableCount -= *resource.Count
+				cpuScaleOutSize -= float64(resource.CPU * *resource.Count)
+
+				plans = append(plans, NewPlan(TiKV, *resource.Count, resource.ResourceType))
+			}
+		}
+	}
+	log.Debug("get heterogeneous scale out plans completed",
+		zap.String("component", component.String()),
+		zap.Any("plans", plans),
+	)
+
+	return plans
+}
+
+func getHeterogeneousScaleInPlans(resourceMap map[string]uint64, component ComponentType) []*Plan {
+	var plans []*Plan
+
+	for resourceType, resourceCount := range resourceMap {
+		if resourceType != homogeneousTiKVResourceType && resourceType != homogeneousTiDBResourceType {
+			plans = append(plans, NewPlan(component, resourceCount-1, resourceType))
+			log.Debug("get heterogeneous scale in plans completed",
```

Comment:
break here to avoid the duplicated code?

Reply:
done